### PR TITLE
Support block mode in tenant cluster

### LINKED ISSUE #64: Support block volumeMode
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What happened**:
The CSI driver only supports filesystem volumeMode. It would be nice to also support block volumeMode.

**What you expected to happen**:
blockVolume mode to create an appropriate PV, and be able to use the block volume in a pod.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt CSI version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
Added support for block mode in tenant cluster.
Added functional test for block mode
Enabled block in k8s e2e suite

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64 

**Special notes for your reviewer**:

**Release note**:

```release-note
Block volumes in tenant cluster
```

### CODE REVIEW

File: pkg/service/node.go [hunk 1]
```diff
@@ -220,29 +221,43 @@ func (n *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+		f, err := os.OpenFile(targetPath, os.O_CREATE, os.FileMode(0644))
+		if err != nil {
+			if !os.IsExist(err) {
+				return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("could not create bind target for block volume %s", targetPath))
+			}
+		} else {
+			_ = f.Close()
+		}
+	} else {
+		err = n.dirMaker.Make(targetPath, 0750)
```
Comment (mhenriks) [inline pkg/service/node.go:249]: doesn't look like we were doing this before so why are we doing it now?

File: pkg/service/node.go [hunk 2]
```diff
@@ -220,29 +221,43 @@ func (n *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
-	klog.Infof("Mounting devicePath %s, on targetPath: %s with FS type: %s",
-		device, targetPath, fsType)
-	err = n.fsMounter.Mount(device.Path, targetPath, fsType, []string{})
+	stagingTarget := req.GetStagingTargetPath()
+	klog.Infof("Staging target path %s", stagingTarget)
+
+	fsType := ""
+	if block {
+		// Make sure target file exists for bind mount
+		f, err := os.OpenFile(targetPath, os.O_CREATE, os.FileMode(0644))
```
Comment (mhenriks) [inline pkg/service/node.go:240]: Why is this necessary?  Will it not get overridden/hidden when bind mount occurs?

File: pkg/service/node.go [hunk 3]
```diff
@@ -270,13 +285,18 @@ func (n *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
+	err := n.mounter.Unmount(targetPath)
 	if err != nil {
 		klog.Infof("Failed to unmount")
 		return nil, err
 	}
 
+	if err = os.RemoveAll(targetPath); err != nil {
+		return nil, fmt.Errorf("remove target path: %w", err)
+	}
+
```
Comment (mhenriks) [inline pkg/service/node.go:299]: related to above comments, if we don't create targetPath then we don't have to remove it

**Review by mhenriks — COMMENTED** [review]
Couple qs about creating target paths

File: pkg/service/node.go (hunk 2 above)
Comment (davidvossel) [reply pkg/service/node.go:240]: also, Is there a reason why opening this file with 0644 is important? or would something like what we have below be fine?

Either way, it might be nice to have a helper function `fileExists` here to reduce some of the branching logic in this function. 

```
func fileExists(path string) (bool, error) {
        _, err := os.Stat(path)
        exists := false

        if err == nil {
                exists = true
        } else if errors.Is(err, os.ErrNotExist) {
                err = nil
        }
        return exists, err
}
```

Comment (awels) [reply pkg/service/node.go:240]: Right but I have to create a file so I can bind mount on top of it. If the file doesn't exist, it will give a message about the mount point not existing. I am bind mounting a single file, not a filesystem.

File: pkg/service/node.go (hunk 1 above)
Comment (awels) [reply pkg/service/node.go:249]: I copied that directly from the original code, we were doing it before. I just moved it into the filesystem case because in the block device case I have to create a file for bind mounting.

File: pkg/service/node.go (hunk 3 above)
Comment (awels) [reply pkg/service/node.go:299]: Right but I do have to create the file so I can bind mount the block device on top of it (needs mount point).

File: e2e/create-pvc_test.go [hunk 4]
```diff
@@ -81,197 +81,224 @@ var _ = Describe("CreatePVC", func() {
 		_ = os.RemoveAll(tmpDir)
 	})
 
-	It("creates a pvc and attaches to pod", Label("pvcCreation"), func() {
-		pvcName := "test-pvc"
-		storageClassName := "kubevirt"
-		pvc := pvcSpec(pvcName, storageClassName, "1Gi")
+	Context("Filesystem volume mode", func() {
+		It("creates a pvc and attaches to pod", Label("pvcCreation"), func() {
```
Comment (davidvossel) [inline e2e/create-pvc_test.go:85]: Many of our existing tests seem to make sense for both block and filesystem pvcs. How practical is it to begin expanding our existing tests to be tables, which execute an entry for filesyste and block.

if this makes sense, I don't think you necessary have to transition all the tests to work for block and filesystem right away. But starting with a small subset could make sense which we expand on later

Comment (awels) [reply e2e/create-pvc_test.go:85]: Should be relatively straight forward, we could pass in a function that builds the pvcs and pods and keep the rest of the logic identical. Note that our tests should test things specific to our provisioner, the k8s e2e test make sure we do the proper things from a k8s perspective. From that perspective block or filesystem should not matter that much to our provisioner.

File: pkg/service/node.go (hunk 1 above)
Comment (mhenriks) [reply pkg/service/node.go:249]: oh I see now

File: pkg/service/node.go (hunk 2 above)
Comment (mhenriks) [reply pkg/service/node.go:240]: huh, so essentially creating a dummy file that gets shadowed by the mount, weird

File: pkg/service/node.go (hunk 3 above)
Comment (mhenriks) [reply pkg/service/node.go:299]: were we removing before?

Comment (awels) [reply pkg/service/node.go:299]: Nope, I believe that was a bug. We were leaving directories behind.

File: pkg/service/node.go [hunk 5]
```diff
@@ -220,29 +221,43 @@ func (n *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
-	req.GetStagingTargetPath()
-	klog.Infof("Staging target path %s", req.GetStagingTargetPath())
-
-	klog.Infof("GetMount() %v", req.VolumeCapability.GetMount())
-	fsType := req.VolumeCapability.GetMount().FsType
-	klog.Infof("Mounting devicePath %s, on targetPath: %s with FS type: %s",
-		device, targetPath, fsType)
-	err = n.fsMounter.Mount(device.Path, targetPath, fsType, []string{})
+	stagingTarget := req.GetStagingTargetPath()
+	klog.Infof("Staging target path %s", stagingTarget)
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/service/node.go:235]: Actually don't need any of that, removing.

File: pkg/service/node.go (hunk 2 above)
Comment (awels) [reply pkg/service/node.go:240]: I can probably do 640 to match the 750 for the directory in case of filesystem volume mode. Let me see if that works.

**Review by davidvossel — COMMENTED** [review]
/approve
